File: projects/portfolio/assets/password_generator.py
import random
import string
from typing import Optional

class PasswordGenerator:
    """Generate secure passwords with customizable options."""
    
    def __init__(self):
        self.lowercase = string.ascii_lowercase
        self.uppercase = string.ascii_uppercase
        self.digits = string.digits
        self.special = string.punctuation
    
    def generate(
        self,
        length: int = 12,
        use_uppercase: bool = True,
        use_lowercase: bool = True,
        use_digits: bool = True,
        use_special: bool = False
    ) -> str:
        """
        Generate a random password.
        
        Args:
            length: Password length (default: 12)
            use_uppercase: Include uppercase letters (default: True)
            use_lowercase: Include lowercase letters (default: True)
            use_digits: Include digits (default: True)
            use_special: Include special characters (default: False)
        
        Returns:
            Generated password string
        """
        if length < 4:
            raise ValueError("Password length must be at least 4")
        
        character_pool = ""
        
        if use_uppercase:
            character_pool += self.uppercase
        if use_lowercase:
            character_pool += self.lowercase
        if use_digits:
            character_pool += self.digits
        if use_special:
            character_pool += self.special
        
        if not character_pool:
            raise ValueError("At least one character type must be selected")
        
        # Generate password
        password = ''.join(random.choice(character_pool) for _ in range(length))
        return password
    
    def generate_multiple(
        self,
        count: int = 5,
        length: int = 12,
        use_uppercase: bool = True,
        use_lowercase: bool = True,
        use_digits: bool = True,
        use_special: bool = False
    ) -> list[str]:
        """Generate multiple passwords."""
        return [
            self.generate(length, use_uppercase, use_lowercase, use_digits, use_special)
            for _ in range(count)
        ]
    
    def get_strength(self, password: str) -> str:
        """
        Evaluate password strength.
        
        Returns:
            'Weak', 'Medium', or 'Strong'
        """
        score = 0
        
        if len(password) >= 12:
            score += 1
        if len(password) >= 16:
            score += 1
        if any(c.isupper() for c in password):
            score += 1
        if any(c.islower() for c in password):
            score += 1
        if any(c.isdigit() for c in password):
            score += 1
        if any(c in self.special for c in password):
            score += 1
        
        if score <= 2:
            return "Weak"
        elif score <= 4:
            return "Medium"
        else:
            return "Strong"


def main():
    """Interactive password generator."""
    pg = PasswordGenerator()
    
    print("=" * 50)
    print("PASSWORD GENERATOR")
    print("=" * 50)
    
    while True:
        print("\nOptions:")
        print("1. Generate single password")
        print("2. Generate multiple passwords")
        print("3. Check password strength")
        print("4. Exit")
        
        choice = input("\nEnter your choice (1-4): ").strip()
        
        if choice == "1":
            try:
                length = int(input("Enter password length (default 12): ") or "12")
                use_upper = input("Include uppercase? (y/n, default y): ").lower() != "n"
                use_lower = input("Include lowercase? (y/n, default y): ").lower() != "n"
                use_digit = input("Include digits? (y/n, default y): ").lower() != "n"
                use_spec = input("Include special characters? (y/n, default n): ").lower() == "y"
                
                password = pg.generate(length, use_upper, use_lower, use_digit, use_spec)
                strength = pg.get_strength(password)
                
                print(f"\nGenerated Password: {password}")
                print(f"Strength: {strength}")
            except ValueError as e:
                print(f"Error: {e}")
        
        elif choice == "2":
            try:
                count = int(input("How many passwords? (default 5): ") or "5")
                length = int(input("Enter password length (default 12): ") or "12")
                use_upper = input("Include uppercase? (y/n, default y): ").lower() != "n"
                use_lower = input("Include lowercase? (y/n, default y): ").lower() != "n"
                use_digit = input("Include digits? (y/n, default y): ").lower() != "n"
                use_spec = input("Include special characters? (y/n, default n): ").lower() == "y"
                
                passwords = pg.generate_multiple(count, length, use_upper, use_lower, use_digit, use_spec)
                
                print(f"\nGenerated {count} passwords:")
                for i, pwd in enumerate(passwords, 1):
                    strength = pg.get_strength(pwd)
                    print(f"{i}. {pwd} ({strength})")
            except ValueError as e:
                print(f"Error: {e}")
        
        elif choice == "3":
            password = input("Enter password to check: ")
            strength = pg.get_strength(password)
            print(f"Strength: {strength}")
        
        elif choice == "4":
            print("Goodbye!")
            break
        
        else:
            print("Invalid choice. Please try again.")


if __name__ == "__main__":
    main()

File: projects/portfolio/js/password-generator.js
function secureRandomInt(maxExclusive) {
	if (!Number.isInteger(maxExclusive) || maxExclusive <= 0) {
		throw new Error('maxExclusive must be a positive integer');
	}

	var uint32Max = 0xffffffff;
	var acceptableRange = uint32Max - (uint32Max % maxExclusive);
	var randomArray = new Uint32Array(1);

	while (true) {
		window.crypto.getRandomValues(randomArray);
		if (randomArray[0] < acceptableRange) {
			return randomArray[0] % maxExclusive;
		}
	}
}

function shuffleString(value) {
	var chars = value.split('');
	for (var i = chars.length - 1; i > 0; i -= 1) {
		var j = secureRandomInt(i + 1);
		var temp = chars[i];
		chars[i] = chars[j];
		chars[j] = temp;
	}
	return chars.join('');
}

function getPasswordStrength(password) {
	var score = 0;

	if (password.length >= 12) {
		score += 1;
	}
	if (password.length >= 16) {
		score += 1;
	}
	if (/[A-Z]/.test(password)) {
		score += 1;
	}
	if (/[a-z]/.test(password)) {
		score += 1;
	}
	if (/\d/.test(password)) {
		score += 1;
	}
	if (/[^A-Za-z0-9]/.test(password)) {
		score += 1;
	}

	if (score <= 2) {
		return 'Weak';
	}
	if (score <= 4) {
		return 'Medium';
	}
	return 'Strong';
}

function buildCharacterSets(options) {
	var sets = [];

	if (options.useUppercase) {
		sets.push('ABCDEFGHIJKLMNOPQRSTUVWXYZ');
	}
	if (options.useLowercase) {
		sets.push('abcdefghijklmnopqrstuvwxyz');
	}
	if (options.useDigits) {
		sets.push('0123456789');
	}
	if (options.useSpecial) {
		sets.push('!\"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~');
	}

	return sets;
}

function generatePassword(length, options) {
	if (length < 4) {
		throw new Error('Password length must be at least 4.');
	}

	var sets = buildCharacterSets(options);
	if (!sets.length) {
		throw new Error('Select at least one character type.');
	}

	if (length < sets.length) {
		throw new Error('Length must be at least ' + sets.length + ' for the selected options.');
	}

	var allCharacters = sets.join('');
	var passwordChars = [];

	sets.forEach(function (set) {
		passwordChars.push(set[secureRandomInt(set.length)]);
	});

	while (passwordChars.length < length) {
		passwordChars.push(allCharacters[secureRandomInt(allCharacters.length)]);
	}

	return shuffleString(passwordChars.join(''));
}

function setupPasswordGeneratorPage() {
	var lengthInput = document.getElementById('password-length');
	if (!lengthInput) {
		return;
	}

	var countInput = document.getElementById('password-count');
	var upperCheckbox = document.getElementById('opt-uppercase');
	var lowerCheckbox = document.getElementById('opt-lowercase');
	var digitsCheckbox = document.getElementById('opt-digits');
	var specialCheckbox = document.getElementById('opt-special');
	var generateSingleButton = document.getElementById('generate-single');
	var generateMultipleButton = document.getElementById('generate-multiple');
	var copyButton = document.getElementById('copy-password');
	var checkStrengthButton = document.getElementById('check-strength');
	var generatedPasswordInput = document.getElementById('generated-password');
	var generatedList = document.getElementById('generated-list');
	var strengthOutput = document.getElementById('password-strength');
	var messageOutput = document.getElementById('generator-message');
	var strengthInput = document.getElementById('strength-input');
	var strengthCheckResult = document.getElementById('strength-check-result');

	function getOptions() {
		return {
			useUppercase: upperCheckbox.checked,
			useLowercase: lowerCheckbox.checked,
			useDigits: digitsCheckbox.checked,
			useSpecial: specialCheckbox.checked
		};
	}

	function showMessage(text, isError) {
		messageOutput.textContent = text;
		messageOutput.classList.toggle('is-error', Boolean(isError));
	}

	generateSingleButton.addEventListener('click', function () {
		try {
			var length = Number(lengthInput.value || '12');
			var password = generatePassword(length, getOptions());
			generatedPasswordInput.value = password;
			strengthOutput.textContent = 'Strength: ' + getPasswordStrength(password);
			showMessage('Password generated.', false);
		} catch (error) {
			showMessage(error.message, true);
		}
	});

	generateMultipleButton.addEventListener('click', function () {
		try {
			var count = Number(countInput.value || '5');
			var length = Number(lengthInput.value || '12');

			if (count < 1 || count > 50) {
				throw new Error('Count must be between 1 and 50.');
			}

			var options = getOptions();
			var lines = [];

			for (var i = 0; i < count; i += 1) {
				var password = generatePassword(length, options);
				lines.push(String(i + 1) + '. ' + password + ' (' + getPasswordStrength(password) + ')');
			}

			generatedList.value = lines.join('\n');
			showMessage('Generated ' + count + ' passwords.', false);
		} catch (error) {
			showMessage(error.message, true);
		}
	});

	copyButton.addEventListener('click', function () {
		var value = generatedPasswordInput.value;
		if (!value) {
			showMessage('Generate a password first.', true);
			return;
		}

		navigator.clipboard.writeText(value).then(function () {
			showMessage('Password copied to clipboard.', false);
		}).catch(function () {
			showMessage('Unable to copy automatically. Please copy manually.', true);
		});
	});

	checkStrengthButton.addEventListener('click', function () {
		var value = strengthInput.value || '';
		if (!value) {
			strengthCheckResult.textContent = 'Strength: -';
			showMessage('Enter a password to check.', true);
			return;
		}

		strengthCheckResult.textContent = 'Strength: ' + getPasswordStrength(value);
		showMessage('Strength checked.', false);
	});
}

document.addEventListener('DOMContentLoaded', function () {
	setupPasswordGeneratorPage();
});
